src/main/java/org/junitpioneer/jupiter/json/JsonFileArgumentsProvider.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2016-2021 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junitpioneer.jupiter.json;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UncheckedIOException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Parameter;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.ArgumentsProvider;
+import org.junit.jupiter.params.support.AnnotationConsumer;
+import org.junit.platform.commons.PreconditionViolationException;
+import org.junitpioneer.internal.PioneerPreconditions;
+import org.junitpioneer.jupiter.CartesianAnnotationConsumer;
+
+/**
+ * The reading of the resources / files is heavily inspired by {@link org.junit.jupiter.params.provider.CsvFileArgumentsProvider}.
+ */
+class JsonFileArgumentsProvider
+		implements ArgumentsProvider, AnnotationConsumer<JsonFileSource>, CartesianAnnotationConsumer<JsonFileSource> {
```

Comment:
Since we're rebuilding Cartesian product tests, I'm not sure whether it makes sense to implement `CartesianAnnotationConsumer<JsonFileSource>`. @Michael1993, what do you think?

Reply:
I've added this because I saw it in the range functionality. I can remove it if it is not needed.

Reply:
Ad hoc decision: For new extensions that can interact with Cartesian tests, let's only implement the new approach. Can you change this? :)